bftengine/src/bftengine/ReadOnlyReplica.cpp

```diff
@@ -0,0 +1,121 @@
+  TimersSingleton::getInstance().cancel(askForCheckpointMsgTimer_);
+  ReplicaForStateTransfer::stop();
+}
+
+void ReadOnlyReplica::onTransferringCompleteImp(int64_t newStateCheckpoint) {
+  lastExecutedSeqNum = newStateCheckpoint;
+  ps_->beginWriteTran();
+  ps_->setLastExecutedSeqNum(lastExecutedSeqNum);
+  ps_->endWriteTran();
+}
+
+void ReadOnlyReplica::onReportAboutInvalidMessage(MessageBase *msg, const char *reason) {
+  LOG_WARN(GL,
+           "Node " << config_.replicaId << " received invalid message from Node " << msg->senderId()
+                   << " type=" << msg->type() << " reason: " << reason);
+}
+void ReadOnlyReplica::sendAskForCheckpointMsg() {
+  LOG_INFO(GL, "sending AskForCheckpointMsg");
+  auto msg = std::make_unique<AskForCheckpointMsg>(config_.replicaId);
+  for (auto id : repsInfo->idsOfPeerReplicas()) send(msg.get(), id);
+}
+
+template <>
+void ReadOnlyReplica::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
+  // metric_received_checkpoints_.Get().Inc();
+  const ReplicaId msgSenderId = msg->senderId();
+  const SeqNum msgSeqNum = msg->seqNumber();
+  const Digest msgDigest = msg->digestOfState();
+  const bool msgIsStable = msg->isStableState();
+
+  LOG_INFO(GL,
+           "Node " << config_.replicaId << " received Checkpoint message from node " << msgSenderId << " for seqNumber "
+                   << msgSeqNum << " (size=" << msg->size() << ", stable=" << (msgIsStable ? "true" : "false")
+                   << ", digestPrefix=" << *((int *)(&msgDigest)) << ")");
+
+  // not relevant
+  if (!msgIsStable || msgSeqNum <= lastExecutedSeqNum) return;
+
+  // previous CheckpointMsg from the same sender
+  auto pos = tableOfStableCheckpoints.find(msgSenderId);
+  if (pos != tableOfStableCheckpoints.end() && pos->second->seqNumber() >= msgSeqNum) return;
+  if (pos != tableOfStableCheckpoints.end()) delete pos->second;
+  CheckpointMsg *x = new CheckpointMsg(msgSenderId, msgSeqNum, msgDigest, msgIsStable);
+  tableOfStableCheckpoints[msgSenderId] = x;
+  LOG_INFO(GL,
+           "Node " << config_.replicaId
+                   << " added stable Checkpoint message to tableOfStableCheckpoints (message from node " << msgSenderId
+                   << " for seqNumber " << msgSeqNum << ")");
+
+  // not enough CheckpointMsg's
+  if ((uint16_t)tableOfStableCheckpoints.size() < config_.fVal + 1) return;
+
+  // check if got enough relevant CheckpointMsgs
+  uint16_t numRelevant = 0;
+  for (auto tableItrator = tableOfStableCheckpoints.begin(); tableItrator != tableOfStableCheckpoints.end();) {
+    if (tableItrator->second->seqNumber() <= lastExecutedSeqNum) {
+      delete tableItrator->second;
+      tableItrator = tableOfStableCheckpoints.erase(tableItrator);
+    } else {
+      numRelevant++;
+      tableItrator++;
+    }
+  }
+  Assert(numRelevant == tableOfStableCheckpoints.size());
+  LOG_INFO(GL, "numRelevant=" << numRelevant);
+
+  // if enough - invoke state transfer
+  if (numRelevant >= config_.fVal + 1) {
```

Comment:
A hardcoded condition? Would add a function that returns (config_.fVal + 1).

Reply:
Copied the code from appropriate place in ReplicaImp

Reply:
i suggest to change to > and remove + 1. Much more intuitive